contracts/BlockRelayProxy.sol

```diff
@@ -73,9 +91,37 @@ contract BlockRelayProxy {
   /// @notice Upgrades the block relay if the current one is upgradeable
   /// @param _newAddress address of the new block relay to upgrade
   function upgradeBlockRelay(address _newAddress) public notIdentical(_newAddress) {
+    // Check if the controller is upgradeable
     require(blockRelayInstance.isUpgradable(msg.sender), "The upgrade has been rejected by the current implementation");
+    // Get last epoch seen by the replaced controller
+    uint256 epoch = blockRelayInstance.getLastEpoch();
+    // Get the length of last epochs seen by the different controllers
+    uint256 n = lastEpochs.length;
+    // If the the last epoch seen by the replaced controller is lower than the one already anotated e.g. 0
+    // just update the already anotated epoch with the new address, ignoring the previously inserted controller
+    // Else, anotate the epoch from which the new controller should start receiving blocks
+    if (epoch < lastEpochs[n-1]) {
+      epochBr[lastEpochs[n-1]] = _newAddress;
+    } else {
+      epochBr[epoch+1] = _newAddress;
+      lastEpochs.push(epoch+1);
+    }
+
+    // Update instance
     blockRelayAddress = _newAddress;
     blockRelayInstance = BlockRelayInterface(_newAddress);
   }
 
+  /// @notice Gets the controller associated with the BR controller corresponding to the epoch provided
+  /// @param _epoch the epoch to work with
+  function getController(uint256 _epoch) public view returns(address _controller) {
+    // Get length of all last epochs seen by controllers
+    uint256 n = lastEpochs.length;
+    // Go backwards until we find the controller having that blockhash
+    for (uint i = 0; i < n; i++) {
```

Comment:
Good choice for the algorithm, maybe be more explicit that this is iterating backwards:

```suggestion
    for (uint i = n; i > 0; i--) {
```

And instead of `n-i-1` use `i-1`.